pkg/fdsrv/fdsrv.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Serves a file descriptor over an AF_UNIX socket when presented with a nonce.
+//
+// You must pass the socket path and nonce to the client via some out-of-band
+// mechanism, such as gRPC or a bash script.
+//
+// Notes:
+// - Uses the unix domain socket abstract namespace
+// - Picks its own path in the abstract namespace for the socket.
+// - Shared FDs are essentially duped, and they point to the same struct file:
+// they share offsets and whatnot.
+//
+// Options:
+// - WithServeOnce: serve once and shuts down (default is forever)
+// - WithTimeout: cancel itself after a timeout (default none)
+//
+// Usage Server:
+//
+//	fds, err := NewFdServer(fd_to_share, "some_nonce", WithServeOnce())
+//	var s path = fds.UdsPath()
+//
+//	// Pass path and some_nonce to the client via an out of band mechanism
+//
+//	fds.Serve(); // Blocks until the server is done
+//	fds.Destroy()
+//
+// Usage Client:
+//
+//	sfd, err := GetSharedFd("uds_path", "some_nonce")
+package fdsrv
+
+import (
+	"errors"
+	"io"
+	"net"
+	"syscall"
+	"time"
+)
+
+type FdServer struct {
```

Comment:
This is a minor thing, I know, but the convention would be to call this a Server, such that the full name would be
fds.Server; otherwise it ends up being fds.FdServer, which is a bit nonstandard.

Reply:
Done